### PR TITLE
Fixes to the mplex implementation

### PR DESCRIPTION
**Pull request by tomaka**
- Right now we ignore data packets of substreams for which we haven't built a `Substream` struct yet, which is bad.
- Added logging.
- Correctly report EOF if the substream was closed.

### CODE REVIEW

**Comment by tomaka** [comment]
Fixed a memory leak with outgoing substreams.

File: mplex/src/lib.rs [hunk 1]
```diff
@@ -265,19 +299,20 @@ where C: AsyncRead + AsyncWrite
...
         let num = try_ready!(next_match(&mut inner, |elem| {
             match elem {
-                codec::Elem::Open { substream_id } => Some(*substream_id),       // TODO: check even/uneven?
+                codec::Elem::Open { substream_id } => Some(*substream_id),
                 _ => None,
             }
         }));
 
         if let Some(num) = num {
-            inner.opened_substreams.insert(num);
```
[the remark this answers is not in the document]
Comment (tomaka) [reply mplex/src/lib.rs]: It's dealt with in `next_match`.

**Comment by tomaka** [comment]
CI fails because of issues fixed by #366.

File: mplex/src/lib.rs [hunk 2]
```diff
@@ -107,77 +111,120 @@ impl<C> Clone for Multiplex<C> {
 
 // Struct shared throughout the implementation.
 struct MultiplexInner<C> {
+    // Errored that happend earlier. Should poison any attempt to use this `MultiplexError`.
+    error: Result<(), IoError>,
     // Underlying stream.
-    inner: Fuse<Framed<C, codec::Codec>>,
+    inner: executor::Spawn<Fuse<Framed<C, codec::Codec>>>,
```
Comment (twittner) [inline mplex/src/lib.rs]: Could you explain a bit what made this change and the introduction of `Notifier` necessary?

Comment (tomaka) [reply mplex/src/lib.rs]: To be honest I thought it was necessary for multiplexing to work correctly, then realized it was not.

**Comment by tomaka** [comment]
I revert the switch to spawning a new task, as it is indeed unnecessary.

**Pull request merged**